// connect-js/js/api.js
var request = require('superagent');
var Q = require('q');
var _ = require('underscore');
var Api;
(function (Api) {
    var QueryResults = (function () {
        function QueryResults(response, cacheKey) {
            if (cacheKey === void 0) { cacheKey = undefined; }
            this.metadata = response.metadata;
            this.results = response.results;
            this.cacheKey = cacheKey;
            if (this.metadata.interval) {
                _.map(this.results, function (intervalResult) {
                    var interval = intervalResult.interval;
                    interval.start = new Date(interval.start);
                    interval.end = new Date(interval.end);
                });
            }
        }
        QueryResults.prototype.selects = function () {
            var results = !this.metadata.interval ? this.results : _.flatten(_.map(this.results, function (result) { return result.results; }));
            return _.difference(_.keys(_.first(results)), this.metadata.groups.concat(['_count']));
        };
        QueryResults.prototype.clone = function () {
            return new QueryResults({
                metadata: _.clone(this.metadata),
                results: JSON.parse(JSON.stringify(this.results))
            }, this.cacheKey);
        };
        return QueryResults;
    })();
    Api.QueryResults = QueryResults;
    var Client = (function () {
        function Client(baseUrl, projectId, apiKey) {
            this._baseUrl = baseUrl;
            this._projectId = projectId;
            this._apiKey = apiKey;
        }
        Client.prototype.query = function (collection, query) {
            var deferred = Q.defer(), queryJson = JSON.stringify(query), url = this._buildUrl('/events/' + collection), get = request.get(url).query({ query: queryJson });
            return this._send(get, function (r) { return new QueryResults(r.body, r.header["etag"]); });
        };
        Client.prototype.pushBatch = function (batches) {
            var url = this._buildUrl('/events'), post = request.post(url).send(batches);
            return this._send(post, function (r) { return r.body; }).deferred.promise;
        };
        Client.prototype.push = function (collection, newEvent) {
            var url = this._buildUrl('/events/' + collection), post = request.post(url).send(newEvent);
            return this._send(post, function (r) { return r.body; }).deferred.promise;
        };
        Client.prototype._send = function (requestToSend, resultsFactory) {
            var deferred = Q.defer();
            requestToSend.set('X-Project-Id', this._projectId).set('X-Api-Key', this._apiKey).end(function (err, res) {
                if (err) {
                    if (!err.status) {
                        err.status = 'NetworkFailure';
                    }
                    deferred.reject(err);
                    return;
                }
                if (!res.ok) {
                    deferred.reject(res.error);
                    return;
                }
                var results = resultsFactory(res);
                deferred.resolve(results);
            });
            return { deferred: deferred, request: requestToSend };
        };
        Client.prototype._buildUrl = function (path) {
            return this._baseUrl + path;
        };
        return Client;
    })();
    Api.Client = Client;
})(Api || (Api = {}));
module.exports = Api;
