// connect-js/js/connect.js
var Api = require('./api');
var Queries = require('./queries/queries');
var _ = require('underscore');
var Connect = (function () {
    function Connect(config) {
        this._config = this.getConfig(config);
        this._client = this._createClient();
    }
    Connect.prototype.push = function (collectionNameOrBatches, eventOrEvents) {
        var hasMultipleEvents = _.isArray(eventOrEvents), hasCollectionName = _.isString(collectionNameOrBatches), isBatch = !hasCollectionName || hasMultipleEvents, batch = isBatch && hasCollectionName ? this._buildBatchFromArray(collectionNameOrBatches, eventOrEvents) : collectionNameOrBatches;
        return isBatch ? this._client.pushBatch(batch) : this._client.push(collectionNameOrBatches, eventOrEvents);
    };
    Connect.prototype.query = function (collection) {
        return new Queries.ConnectQuery(this._client, collection);
    };
    Connect.prototype.getConfig = function (config) {
        return {
            baseUrl: config.baseUrl || 'https://api.getconnect.io',
            projectId: config.projectId,
            apiKey: config.apiKey
        };
    };
    Connect.prototype._createClient = function () {
        return new Api.Client(this._config.baseUrl, this._config.projectId, this._config.apiKey);
    };
    Connect.prototype._buildBatchFromArray = function (collection, events) {
        var batch = {};
        batch[collection] = events;
        return batch;
    };
    Connect.QueryResults = Api.QueryResults;
    return Connect;
})();
module.exports = Connect;
