// connect-js/js/queries/query-builder.js
var QueryBuilder = (function () {
    function QueryBuilder() {
    }
    QueryBuilder.prototype.build = function (selects, filters, groups, timeframe, interval, timezone, customOptions) {
        var query = {
            select: selects || {}
        };
        if (filters.length > 0) {
            query['filter'] = this._buildFilter(filters);
        }
        if (groups.length > 0) {
            query['groupBy'] = groups;
        }
        if (timeframe) {
            query['timeframe'] = timeframe;
        }
        if (interval) {
            query['interval'] = interval;
        }
        if (timezone) {
            query['timezone'] = timezone;
        }
        if (customOptions) {
            for (var name in customOptions)
                query[name] = customOptions[name];
        }
        return query;
    };
    QueryBuilder.prototype._buildFilter = function (filters) {
        var queryFilter = {};
        filters.forEach(function (filter) {
            var fieldFilters = queryFilter[filter.field];
            if (!fieldFilters) {
                fieldFilters = {};
                queryFilter[filter.field] = fieldFilters;
            }
            fieldFilters[filter.operator] = filter.value;
        });
        return queryFilter;
    };
    return QueryBuilder;
})();
module.exports = QueryBuilder;
